### PR TITLE
Cleanup ParentFieldMapper

### PR DESCRIPTION
**Pull request by martijnvg**
The only purpose of the `ParentFieldMapper` is to make sure that the `has_child` query, `has_parent` query and `children` aggregation work. With this in mind the `ParentFieldMapper` should be cleaned up:
- Stop storing stored fields and indexed fields. The _parent field's only purpose is to support joins between parent and child type and only storing doc value fields is sufficient.
- In the mapping the parent field mapper is now known under '{parent}#{child}' key, because this is the field the parent/child join uses too.
- Added new sub fetch phase to lookup that _parent field from doc values field if that is requested. (before this was fetched from stored _parent field)
- Removed the ability to query directly on `_parent` in the query dsl. Instead the `{parent}#{child}` field should be used. Under the hood a doc values query is used instead of a term query, because only doc values fields are stored now.
- Also in aggregations `_parent` field can't be used any more and `{parent}#{child}` field name should be used instead to aggregate directly on the `_parent` join field.
- Remove remaining 1.x bwc logic.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/mapper/internal/ParentFieldMapper.java [hunk 1]
```diff
@@ -88,17 +79,13 @@
...
 
         private final MappedFieldType parentJoinFieldType = Defaults.JOIN_FIELD_TYPE.clone();
 
-        private final MappedFieldType childJoinFieldType = Defaults.JOIN_FIELD_TYPE.clone();
-
         public Builder(String documentType) {
-            super(Defaults.NAME, Defaults.FIELD_TYPE, Defaults.FIELD_TYPE);
-            this.indexName = name;
+            super(Defaults.NAME, new ParentFieldType(Defaults.JOIN_FIELD_TYPE, documentType), Defaults.JOIN_FIELD_TYPE);
+            docValues(true);
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/index/mapper/internal/ParentFieldMapper.java]: Why is this needed when doc values is set in the JOIN_FIELD_TYPE default?

Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/index/mapper/internal/ParentFieldMapper.java]: this is because `FieldMapper.Builder#docValuesSet` boolean field. It needs to be set to true, otherwise `#setupFieldType(...)` doesn't enable doc values on fieldType.

Maybe instead of invoking this `docValues(...)` method, the `FieldMapper.Builder` constructor should initialise `docValuesSet` value to whether doc values has been enabled on `fieldType` parameter?

Comment (rjernst) [inline core/src/main/java/org/elasticsearch/index/mapper/internal/ParentFieldMapper.java]: Yes, I think the builder should operate on the field type.

**Comment by rjernst** [comment]
One problem I see that is pre-existing, but that we should fix, is that this `{parent}#{child}` field is essentially a subfield. It should really be returned as a subfield from the mapper, so that doing a PUT mapping will reject creating a mapper with the same name.

**Comment by rjernst** [comment]
Just to be clear, by return that as a subfield I mean that it should have its own mapper, and that should be returned by overriding `iterator()` from `FieldMapper` in `ParentFieldMapper`.

**Comment by martijnvg** [comment]
Thanks for checking this out @rjernst. I've updated this PR.

File: core/src/main/java/org/elasticsearch/index/get/ShardGetService.java [hunk 2]
```diff
@@ -350,6 +355,16 @@ private GetResult innerGetLoadFromStoredFields(String type, String id, String[]
             }
         }
 
+        if (docMapper.parentFieldMapper().active()) {
+            String parentId = ParentFieldSubFetchPhase.getParentId(docMapper.parentFieldMapper(), docIdAndVersion.context.reader(), docIdAndVersion.docId);
+            if (parentId != null) {
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/index/get/ShardGetService.java]: What does it mean if parent id is not found?

**Comment by rjernst** [comment]
The change looks good, but the one thing that still bugs me is a user having to know the special syntax of a field we auto create, `{parent}#{child}`. Should this be a query that takes both parent and child so the user does not need to know how we build this field internally?

**Comment by martijnvg** [comment]
@rjernst agreed would be great if just `_parent:{value}` can be used. The problem is that in the query dsl we can't know for sure what child type a user is referring to. There can several parent/child relations defined in an index. Currently the non lenient syntax in master/2.x is: `_parent:{parent_type}#{id}`. I can bring back the lenient syntax that just tries to find any child doc matching with a parent id value?

**Comment by rjernst** [comment]
I did not mean to suggest adding back the previous behavior. I meant having a query in the dsl that does the joining of the values, so that a user doesn't need to know how the values are encoded. Eg something like:

```
{"query": {"parentquery":{"type": "footype", "id": "myparentid"}}}
```

Fill in "parentquery" with a name that actually makes sense, but I think that demonstrates the idea.

File: core/src/main/java/org/elasticsearch/index/get/ShardGetService.java (hunk 2 above)
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/index/get/ShardGetService.java]: this shouldn't happen. If the parent field is active then a parent id has to be specified, so upon retrieval it should be there.  I'll add an assert here for this.

**Comment by martijnvg** [comment]
+1 Adding such a query makes it easer to query child docs by parent id.

**Comment by martijnvg** [comment]
I've updated the PR and added a new `parent_id` query that selects the right field to be used.

**Comment by rjernst** [comment]
Thanks @martijnvg! LGTM.
